.github/copilot-instructions.md

```diff
@@ -0,0 +1,413 @@
+tar -xf shellcheck-latest.linux.x86_64.tar.xz
+sudo cp shellcheck-latest/shellcheck /usr/local/bin/
+```
+
+### Usage
+
+#### Lint Main Files
+```bash
+shellcheck -s bash nvm.sh
+shellcheck -s bash install.sh  
+shellcheck -s bash nvm-exec
+shellcheck -s bash bash_completion
+```
+
+#### Lint Across Shell Types
+```bash
+shellcheck -s sh nvm.sh      # POSIX sh
+shellcheck -s bash nvm.sh    # Bash extensions
+shellcheck -s dash nvm.sh    # Dash compatibility
+shellcheck -s ksh nvm.sh     # Ksh compatibility
+```
+
+#### Common shellcheck Directives in nvm
+- `# shellcheck disable=SC2039` - Allow bash extensions in POSIX mode
+- `# shellcheck disable=SC2016` - Allow literal `$` in single quotes
+- `# shellcheck disable=SC2001` - Allow sed usage instead of parameter expansion
+- `# shellcheck disable=SC3043` - Allow `local` keyword (bash extension)
+
+### Fixing shellcheck Issues
+1. **Quoting**: Always quote variables: `"${VAR}"` instead of `$VAR`
+2. **POSIX compliance**: Avoid bash-specific features in portable sections
+3. **Array usage**: Use `set --` for positional parameters instead of arrays, which are not supported in POSIX
+4. **Local variables**: Declared with `local FOO` and then initialized on the next line (the latter is for ksh support)
+
+## Development Best Practices
+
+### Code Style
+- Use 2-space indentation
+- Follow POSIX shell guidelines for portability
+- Prefix internal functions with `nvm_`
+- Use `nvm_echo` instead of `echo` for output
+- Use `nvm_err` for error messages
+
+### Compatibility
+- Test changes across all supported shells
+- Avoid bash-specific features in core functionality
+- Use `nvm_is_zsh` to check when zsh-specific behavior is needed
+- Mock external dependencies in tests
+
+### Performance
+- Cache expensive operations (like remote version lists)
+- Use local variables to avoid scope pollution
+- Minimize subprocess calls where possible
+- Implement lazy loading for optional features
+
+### Debugging
+- Use `nvm debug` command for environment information
+- Enable verbose output with `set -x` during development
+- Test with `NVM_DEBUG=1` environment variable
+- Check `$NVM_DIR/.cache` for cached data issues
+
+## Common Gotchas
+
+1. **PATH modification**: nvm modifies PATH extensively; be careful with restoration
+2. **Shell sourcing**: nvm must be sourced, not executed as a script
+3. **Version resolution**: Aliases, partial versions, and special keywords interact complexly
+4. **Platform differences**: Handle differences between Linux, macOS, and other Unix systems
+5. **Network dependencies**: Many operations require internet access for version lists
+6. **Concurrent access**: Multiple shells can conflict when installing versions simultaneously
+
+## Windows Support
+
+nvm works on Windows via several compatibility layers:
+
+### WSL2 (Windows Subsystem for Linux)
+- Full nvm functionality available
+- Install WSL2 with Ubuntu or other Linux distribution
```

Comment:
please provide instructions here to ensure the contributor is on the latest version of wsl2, not wsl1 which won't work properly (a link to a working and up-to-date guide is sufficient)

(i see the instructions in a later section, but let's interleave them here in these sections instead)